Mathlib/RingTheory/Polynomial/Hilbert.lean

```diff
@@ -0,0 +1,105 @@
+  _ = (n - k + d).choose d := by
+    rw [ascPochhammer_nat_eq_natCast_ascFactorial];
+    field_simp [ascFactorial_eq_factorial_mul_choose]
+
+variable {F}
+
+/--
+`Polynomial.hilbert p 0 = 0`; for any `d : ℕ`, `Polynomial.hilbert p (d + 1)` is
+defined as `∑ i in p.support, (p.coeff i) • Polynomial.preHilbert F d i`. If `M` is
+a graded module whose Poincaré series can be written as `p(X)/(1 - X)ᵈ` for some
+`p : ℚ[X]` with integer coefficients, then `Polynomial.hilbert p d` is the Hilbert
+polynomial of `M`. See also `Polynomial.coeff_mul_invOneSubPow_eq_hilbert_eval`,
+which says that `PowerSeries.coeff F n (p * (PowerSeries.invOneSubPow F d))` is
+equal to `(Polynomial.hilbert p d).eval (n : F)` for any large enough `n : ℕ`.
+-/
+noncomputable def hilbert (p : F[X]) : (d : ℕ) → F[X]
+  | 0 => 0
+  | d + 1 => ∑ i in p.support, (p.coeff i) • preHilbert F d i
+
+variable (F) in
+lemma hilbert_zero (d : ℕ) : hilbert (0 : F[X]) d = 0 := by
+  delta hilbert; induction d with
+  | zero => simp only
+  | succ d _ => simp only [coeff_zero, zero_smul, Finset.sum_const_zero]
+
+/--
+The key property of Hilbert polynomials. If `F` is a field with characteristic `0`, `p : F[X]` and
+`d : ℕ`, then for any large enough `n : ℕ`, `(Polynomial.hilbert p d).eval (n : F)` is equal to the
+coefficient of `Xⁿ` in the power series expansion of `p/(1 - X)ᵈ`.
+-/
+theorem coeff_mul_invOneSubPow_eq_hilbert_eval
+    [CharZero F] (p : F[X]) (d n : ℕ) (hn : p.natDegree < n) :
+    PowerSeries.coeff F n (p * (invOneSubPow F d)) = (hilbert p d).eval (n : F) := by
+  delta hilbert; induction d with
+  | zero => simp only [invOneSubPow_zero, Units.val_one, mul_one, coeff_coe, eval_zero]
+            exact coeff_eq_zero_of_natDegree_lt hn
+  | succ d hd =>
+      simp only [eval_finset_sum, eval_smul, smul_eq_mul]; rw [← Finset.sum_coe_sort]
+      simp_rw [show (i : p.support) → eval ↑n (preHilbert F d ↑i) = (n + d - ↑i).choose d by
+        intro i; rw [preHilbert_eq_choose_sub_add _ _ _ _ <| le_trans (le_natDegree_of_ne_zero
+        <| mem_support_iff.1 i.2) (le_of_lt hn)]; rw [Nat.sub_add_comm];
+        exact le_trans (le_natDegree_of_ne_zero <| mem_support_iff.1 i.2) (le_of_lt hn)]
+      rw [Finset.sum_coe_sort _ (fun x => (p.coeff ↑x) * (_ + d - ↑x).choose _),
+        PowerSeries.coeff_mul, Finset.Nat.sum_antidiagonal_eq_sum_range_succ_mk,
+        invOneSubPow_val_eq_mk_sub_one_add_choose_of_pos _ _ (zero_lt_succ d)]
+      simp only [coeff_coe, coeff_mk]
+      exact Eq.symm <| Finset.sum_subset_zero_on_sdiff (fun s hs => Finset.mem_range_succ_iff.mpr
+        <| le_trans (le_natDegree_of_ne_zero <| mem_support_iff.1 hs) (le_of_lt hn)) (fun x hx => by
+        simp only [Finset.mem_sdiff, mem_support_iff, not_not] at hx; rw [hx.2, zero_mul])
+        (fun x hx => by rw [add_comm, Nat.add_sub_assoc <| le_trans (le_natDegree_of_ne_zero <|
+        mem_support_iff.1 hx) (le_of_lt hn), succ_eq_add_one, add_tsub_cancel_right])
```

Comment:
Oh come on, have mercy on the people who will have to refactor the proof when it breaks. There is no reason to use any semicolons, bytes are cheap. I really don't like `rw [show something with a three line proof in the middle of the rewrite, which uses more rewrites]`, just `have h :=` the thing you want to rewrite, prove it, and `rw [h]`. Similarly surely `exact Eq.symm <| (some lemma with nontrivial inputs)` would just be better written as `apply`ing the lemma and then solving all the intermediate goals. This sort of proof will randomly break one day and then it will be the job of the poor person who has to fix it to unravel this mess.

Reply:
[NB Fangming and I just had an offline discussion about how to tidy this proof up; he has a much more comprehensible version incoming]